clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java

```diff
@@ -727,6 +731,20 @@ private void checkGroupRemoteAssignor() {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private void checkPartitionAssigmentStrategy() {
+        List<Object> assignmentStrategies = (List<Object>) get(PARTITION_ASSIGNMENT_STRATEGY_CONFIG);
+        if (getString(GROUP_PROTOCOL_CONFIG).equalsIgnoreCase(GroupProtocol.CONSUMER.name())) {
+            if (PARTITION_ASSIGNOR_DEFAULT_VALUE.size() != assignmentStrategies.size() ||
+                    !PARTITION_ASSIGNOR_DEFAULT_VALUE.stream()
+                            .allMatch(clz -> assignmentStrategies.stream()
+                                    .anyMatch(obj -> clz.equals(obj) || clz.getName().equals(obj)))
+            ) {
+                throw new ConfigException(PARTITION_ASSIGNMENT_STRATEGY_CONFIG + " cannot be set when " + GROUP_PROTOCOL_CONFIG + "=" + GroupProtocol.CONSUMER.name());
+            }
+        }
+    }
+
```

Comment:
I think the core check could be much simpler 🤔

I used the `originals()` method to check the values the user provided before any defaults were applied.

```suggestion
    private void checkPartitionAssignmentStrategy() {
        if (!getString(GROUP_PROTOCOL_CONFIG).equalsIgnoreCase(GroupProtocol.CONSUMER.name()))
            return;

        Object config = originals().get(PARTITION_ASSIGNMENT_STRATEGY_CONFIG);

        if (config != null && !Utils.isBlank(config.toString())) {
            throw new ConfigException(PARTITION_ASSIGNMENT_STRATEGY_CONFIG + " cannot be set when " + GROUP_PROTOCOL_CONFIG + "=" + GroupProtocol.CONSUMER.name());
        }
    }

```

Is there anything missing in the above?

Reply:
Hello @kirktrue, Thanks for your comments, I think we have two way to discuss this issue
If we want to block any input of `partition.assignment.strategy`, your method is more simplify and better, however if we allow user to input the default value of `partition.assignment.strategy`, we use the origin method is more easier to check.

Reply:
I would say we want to throw if there is any value defined by the user for this property (it's truly deprecated with the new protocol), so @kirktrue 's suggestion of working with the originals seems effective and simpler.

Reply:
+1 to @kirktrue @lianetm that checking the user-defined values makes sense to me.

For another, `ConsumerConfig`'s constructor will log all configs, so it would be nice to set the config `partition.assignment.strategy` to null when the protocol is `CONSUMER`.  This behavior is similar to `maybeOverrideEnableAutoCommit` which can avoid printing "weird" consumer configs in log file.

Reply:
Do you refer to the log for unused? I did think about that, but then checked and noticed that it only logs based on the originals, not the defaults , so seems we shouldn't need to override to avoid the log https://github.com/apache/kafka/blob/005155ba5cad313d22fb84efc9959dd3fd484bf3/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L225

Reply:
> Do you refer to the log for unused?

I think about `logAll()` 
https://github.com/apache/kafka/blob/005155ba5cad313d22fb84efc9959dd3fd484bf3/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L121

If we don't remove the default value of  `partition.assignment.strategy`, it will print following log.
```
 [2024-09-04 19:45:40,533] INFO ConsumerConfig values: 
        ...
	group.protocol = CONSUMER
        ...
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor, class org.apache.kafka.clients.consumer.CooperativeStickyAssignor]
```

That is not a big issue, but the output could misdirect users to assume "those" configs are valid.

Reply:
It make sense to me. +1 to @chia7712, the config log should not misdirect user.

Reply:
totally agree it will be confusing on the `logAll`, +1 to remove

Reply:
@m1a2st I think this comment can be addressed in a follow-up (https://issues.apache.org/jira/browse/KAFKA-17821). Could you please revert the related changes? Sorry for complicating this PR.